File: Data Structures & Algorithms/search-2d-matrix/submission-0.cs
public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        // first we search the row based on the first entry of each row
        // we also have to make sure it's not smaller than the first entry 
        // and not larger than the last
        // after finding the row, we find the exact element using binary search

        int[] joined = matrix[0];
        for(int i = 1; i < matrix.Length; i++)
        {
            joined = joined.Concat(matrix[i]).ToArray();
        }

        if(target < joined[0] || target > joined.Last())
        {
            return false;
        }
        else 
        {
            var lo = 0;
            var hi = joined.Length - 1;
            while (lo <= hi) 
            {
                var mid = (int)Math.Ceiling((float)((lo + hi) /2));
                if (joined[mid] == target)
                {
                    return true;
                }
                else if (joined[mid] < target) 
                {
                    lo = mid + 1;
                }
                else if (joined[mid] > target)
                {
                    hi = mid - 1;
                }
            }
            return false;
        }
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        minimum = nums[0]
        lo = 0
        hi = len(nums) - 1
        offset = 0
        while lo < hi:
            offset = (lo + hi) // 2
            if nums[offset] < nums[hi]:
                minimum = nums[offset]
                hi = offset
            else:
                lo = offset + 1
        offset = lo
        print(offset)

        lo = 0
        hi = len(nums) - 1
        while lo <= hi:
            logical_mid = (lo + hi) // 2
            mid = (logical_mid + offset) % len(nums)
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                lo = logical_mid + 1
            else:
                hi = logical_mid -1

        return -1